### ISSUE TITLE
Azure builds failing with SEGV

### ISSUE THREAD

**Issue by ndevenish** [cctbx/dxtbx#383]
Azure builds are currently failing with a SEGV. This is reproduceable on an environment that uses `cctbx_base` from conda instead of building from scratch (so e.g. doesn't show on Jenkins).

So far, I'm struggling with the fact that the libtbx-module-installed-splitting is somewhat a prototype, and:
- It's impossible to build and test in the same environment (activating `setpaths.sh` means `make` fails)
- calling `libtbx/configure.py` to set up a debug environment fails to build
- Even bypassing this to set flags directly in `build/libtbx_env` fails - It's apparently _intentionally_ impossible to build with different flags than the parent environment (so utterly impossible to build e.g. debug symbols into a library in the libtbx build system against the compiled base)

**Comment by Anthchirp** [comment]
The issues with building on top of a prebuilt cctbx are probably of interest to @bkpoon

**Comment by ndevenish** [comment]
re: getting a debug build: `rm lib/dxtbx* && rm -rf dxtbx/ && make | grep x86_64-conda-linux-gnu-c++ | tail -n +2 | sed '/$/ -g -O0/' > cc.txt; . cc.txt` mostly works, with needing to edit the command list to remove the rogue `$(` and `)$` that apparently snuck into the linking step.

**Comment by ndevenish** [comment]
More random info of my current state of tracking down:

Reproducing `Dockerfile`
```Dockerfile
FROM ubuntu:18.04
RUN apt-get update && apt-get install -y git python3 make
WORKDIR /opt
RUN git clone https://github.com/cctbx/dxtbx modules/dxtbx && \
    git clone https://github.com/dials/annlib modules/annlib && \
    git clone https://github.com/dials/annlib_adaptbx modules/annlib_adaptbx
RUN python3 modules/dxtbx/.azure-pipelines/bootstrap.py base --clean --python 3.9
RUN python3 modules/dxtbx/.azure-pipelines/bootstrap.py update
RUN bash -c ". conda_base/bin/activate && \
    mkdir build && \
    cd build && \
    libtbx.configure dxtbx cbflib_adaptbx && \
    make"
RUN bash -c ". conda_base/bin/activate && . build/setpaths.sh && \
    libtbx.python -c '\
from scitbx.array_family import flex;\
from dxtbx.model import Scan; \
Scan((1,1), (18.0,0.5), flex.double([1.0]), flex.double([0]), 0, True) \
'"
```
Stack frame:
```
* thread #1, name = 'python3.9', stop reason = signal SIGSEGV: invalid address (fault address: 0x12)
    frame #0: 0x00007fffeb3d9a86 dxtbx_model_ext.so`scitbx::af::shared_plain<double>::m_dispose(this=0x00007fffffffaa70) at shared_plain.h:375:36
   372 	        if (m_handle->use_count == 0) {
   373 	          clear();
   374 	          if (m_handle->weak_count == 0) delete m_handle;
-> 375 	          else m_handle->deallocate();
   376 	        }
   377 	      }
   378
```

Trace doesn't seem to get to the actual code:
```
* thread #1, name = 'python3.9', stop reason = signal SIGSEGV: invalid address (fault address: 0x12)
  * frame #0: 0x00007fffeb3d9a86 dxtbx_model_ext.so`scitbx::af::shared_plain<double>::m_dispose(this=0x00007fffffffaa50) at shared_plain.h:375:36
    frame #1: 0x00007fffeb3d85f6 dxtbx_model_ext.so`scitbx::af::shared_plain<double>::~shared_plain(this=0x00007fffffffaa50) at shared_plain.h:227:9
    frame #2: 0x00007fffeb3d6f98 dxtbx_model_ext.so`scitbx::af::shared<double>::~shared(this=0x00007fffffffaa50) at shared.h:11:9
    frame #3: 0x00007fffeb3e86f2 dxtbx_model_ext.so`void boost::python::detail::value_destroyer<false>::execute<scitbx::af::shared<double> >(p=0x00007fffffffaa50) at destroy.hpp:20:9
    frame #4: 0x00007fffeb50120b dxtbx_model_ext.so`void boost::python::detail::destroy_referent_impl<scitbx::af::shared<double> const>(p=0x00007fffffffaa50, (null)=0x0000000000000000)()) at destroy.hpp:52:15
    frame #5: 0x00007fffeb500a89 dxtbx_model_ext.so`void boost::python::detail::destroy_referent<scitbx::af::shared<double> const&>(p=0x00007fffffffaa50, (null)=0x0000000000000000)()) at destroy.hpp:58:26
    frame #6: 0x00007fffeb4fe015 dxtbx_model_ext.so`boost::python::converter::rvalue_from_python_data<scitbx::af::shared<double> const&>::~rvalue_from_python_data(this=0x00007fffffffaa40) at rvalue_from_python_data.hpp:135:51
    frame #7: 0x00007fffeb4f7fa8 dxtbx_model_ext.so`boost::python::converter::arg_rvalue_from_python<scitbx::af::shared<double> const&>::~arg_rvalue_from_python(this=0x00007fffffffaa40) at arg_from_python.hpp:106:8
    frame #8: 0x00007fffeb4f7fc4 dxtbx_model_ext.so`boost::python::arg_from_python<scitbx::af::shared<double> const&>::~arg_from_python(this=0x00007fffffffaa40) at arg_from_python.hpp:18:8
    frame #9: 0x00007fffeb4f9636 dxtbx_model_ext.so`boost::python::detail::caller_arity<6u>::impl<dxtbx::model::Scan* (this=0x00005555561044e8, args_=('0x7fffec44a5f8', (1, 1), (18.0, 0.5), '0x7fffec44a610', '0x7fffec44a618', 0, True), (null)='0x7fffffffa948')(scitbx::vec2<int>, scitbx::vec2<double>, scitbx::af::shared<double> const&, scitbx::af::shared<double> const&, int, bool), boost::python::detail::constructor_policy<boost::python::default_call_policies>, boost::mpl::vector7<dxtbx::model::Scan*, scitbx::vec2<int>, scitbx::vec2<double>, scitbx::af::shared<double> const&, scitbx::af::shared<double> const&, int, bool> >::operator()(_object*, _object*) at local.hpp:43:9
    frame #10: 0x00007fffeb4f46c3 dxtbx_model_ext.so`boost::python::objects::signature_py_function_impl<boost::python::detail::caller<dxtbx::model::Scan* (*)(scitbx::vec2<int>, scitbx::vec2<double>, scitbx::af::shared<double> const&, scitbx::af::shared<double> const&, int, bool), boost::python::detail::constructor_policy<boost::python::default_call_policies>, boost::mpl::vector7<dxtbx::model::Scan*, scitbx::vec2<int>, scitbx::vec2<double>, scitbx::af::shared<double> const&, scitbx::af::shared<double> const&, int, bool> >, boost::mpl::v_item<void, boost::mpl::v_item<boost::python::api::object, boost::mpl::v_mask<boost::mpl::vector7<dxtbx::model::Scan*, scitbx::vec2<int>, scitbx::vec2<double>, scitbx::af::shared<double> const&, scitbx::af::shared<double> const&, int, bool>, 1>, 1>, 1> >::operator(this=0x00005555561044e0, args=('0x7fffec44a5f8', (1, 1), (18.0, 0.5), '0x7fffec44a610', '0x7fffec44a618', 0, True), kw='0x7fffffffaaa8')(_object*, _object*) at py_function.hpp:64:33
    frame #11: 0x00007ffff03eaecd libboost_python39.so.1.74.0`boost::python::objects::function::call(_object*, _object*) const + 749
    frame #12: 0x00007ffff03eb0d9 libboost_python39.so.1.74.0`boost::detail::function::void_function_ref_invoker0<boost::python::objects::(anonymous namespace)::bind_return, void>::invoke(boost::detail::function::function_buffer&) + 25
    frame #13: 0x00007ffff03efe23 libboost_python39.so.1.74.0`boost::python::handle_exception_impl(boost::function0<void>) + 115
    frame #14: 0x00007ffff03e8093 libboost_python39.so.1.74.0`function_call + 99
    frame #15: 0x0000555555699bb6 python3.9`_PyObject_MakeTpCall at call.c:191:18
```

**Comment by ndevenish** [comment]
I've discovered the issue. When libtbx is configured as a split-install split-development environment, apparently it's required to check out a second copy of `cctbx_project`. However, nothing verifies that this is the same version - in this case https://github.com/cctbx/cctbx_project/commit/ab8949cd765dd31ef8749a5b5950459ffdf35c36 had been included on the `stable` branch. This meant that dxtbx was compiled with a different memory layout of the classes to the ones used within cctbx, so passing them from scitbx to dxtbx broke as dxtbx was expecting a polymorphic class.

TLDR; the build system is broken.

**Comment by bkpoon** [comment]
Another way is to delete everything in `cctbx_project` except for `cbflib_adaptbx` and `libtbx`. You only need `cbflib_adaptbx`. And `libtbx` is only needed for running `libtbx.refresh`, but I can probably update the logic so that it is not needed. The `SConstruct` file will find the necessary headers in the environment. I'm not sure why the build was preferring the ones in `cctbx_project` first.

The compilation flags can be modified, but it requires editing the environment file in `${CONDA_PREFIX}/share/cctbx`. This can be done by activating the environment, but not sourcing any of the `setpaths` files. Otherwise, the environment in the `build` directory is modified.
```
import libtbx.load_env
e = libtbx.env
e.build_options.env_cxxflags = '-O3'
e.pickle()
```
